### PR TITLE
List consumer group offsets

### PR DESCRIPTION
**Pull request by PratRanj07**
Created ListConsumerGroupOffsets API with the name of fetchOffsets and wrote examples and test cases for it.

### CODE REVIEW

File: test/promisified/admin/list_consumer_group_offsets.spec.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+  it("should timeout when fetching offsets", async () => {
+    await admin.connect();
+
+    await expect(
+      admin.fetchOffsets({ groupId, topic: topicName, timeout: 0 })
+    ).rejects.toHaveProperty("code", ErrorCodes.ERR__TIMED_OUT);
+  });
+
+  test("should return correct offset after consuming messages", async () => {
```
Comment (milindl) [inline test/promisified/admin/list_consumer_group_offsets.spec.js:48]: - add a test with `topics` unset or null too
- add a test with multiple topics each with more than 1 partition so the result format can be verified

File: test/promisified/admin/list_consumer_group_offsets.spec.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
+jest.setTimeout(30000);
```
Comment (milindl) [inline test/promisified/admin/list_consumer_group_offsets.spec.js:1]: rename this file to be same as the function name. Let's call it 'list consumer groups' only when we actually need to call into librdkafka. Otherwise it gets confusing to mix the names.

File: types/kafkajs.d.ts [hunk 3]
```diff
@@ -259,6 +259,10 @@ export interface PartitionOffset {
   offset: string
 }
 
+export type FetchOffsetsPartition = PartitionOffset & { metadata: string | null }
+
+export type TopicInput = string | { topic: string; partitions: number[] }
```
Comment (milindl) [inline types/kafkajs.d.ts:264]: To be consistent with how we have done things before, define these types in rdkafka.d.ts, and import them here. Re-export from this file. (See the top of this file for examples, one such type is `GroupOverview`)

File: types/rdkafka.d.ts [hunk 4]
```diff
@@ -11,7 +11,7 @@ import {
 
 export * from './config';
 export * from './errors';
-import { Kafka } from './kafkajs';
+import { FetchOffsetsPartition, Kafka, TopicInput } from './kafkajs';
```
Comment (milindl) [inline types/rdkafka.d.ts:14]: Remove this line entirely after following the comment from kafkajs.d.ts (`Kafka` was unused anyway)

**Review by milindl — CHANGES_REQUESTED** [review]
Good work @PratRanj07  to implement this. 

I have done a first round of review.
Things still pending to be reviewed by me:
1. example added
2. run leak checks with valgrind 

and of course any changes you make due to the comments.

**Comment by EricMCornelius** [comment]
Any ETA on this?  You guys are promoting this client as a KafkaJS replacement while lacking some pretty basic functionality on the admin side here...

**Comment by EricMCornelius** [comment]
https://github.com/confluentinc/confluent-kafka-javascript/blob/master/MIGRATION.md#admin-client

At the very least, maybe keep major omissions up to date in any migration guides?

File: types/rdkafka.d.ts [hunk 5]
```diff
@@ -438,6 +442,12 @@ export interface IAdminClient {
     deleteGroups(groupIds: string[],
         options?: { timeout?: number },
         cb?: (err: LibrdKafkaError, result: DeleteGroupsResult[]) => any): void;
+    fetchOffsets(options: { groupId: string,
```
Comment (emasab) [inline types/rdkafka.d.ts:445]: for the `node-rdkafka` API, it's a new method, so we should have the same signature and name as `listConsumerGroupOffsets`, allowing multiple groups. In future, we'll deprecate `fetchOffsets` and have a new method named `listConsumerGroupOffsets` in KafkaJS API too.

File: src/admin.cc [hunk 6]
```diff
@@ -666,6 +667,91 @@ Baton AdminClient::DeleteGroups(rd_kafka_DeleteGroup_t **group_list,
...
+    }
+
+    const rd_kafka_ListConsumerGroupOffsets_result_t *result =
+        rd_kafka_event_ListConsumerGroupOffsets_result(*event_response);
+
+    size_t result_cnt;
+    const rd_kafka_group_result_t **results =
+        rd_kafka_ListConsumerGroupOffsets_result_groups(result, &result_cnt);
+
+    // Change the type of the 'error' pointer to 'const rd_kafka_error_t *'
```
Comment (milindl) [inline src/admin.cc:742]: Let's move this processing outside this function and let the caller process this. This sort of works for now, but when/if we add the capacity for multiple groups, it won't.

Let's handle only top level errors here, and move this group-specific errors out.

Comment (milindl) [reply src/admin.cc:742]: So with the change that Emanuele suggested, you will need to change the FromFetchOffsetsResult function as well, since the return type signature will be different for listConsumerGroupOffsets.

However, the kafkajs-compatible signature remains the same, so you'll need to convert it in `fetchOffsets` in _admin.js

File: test/promisified/admin/fetch_offsets.spec.js [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+          messagesConsumed.push(message); // Populate messagesConsumed
+          if (messagesConsumed.length === 5) {
+            await consumer.commitOffsets([
+              {
+                topic,
+                partition,
+                offset: (parseInt(message.offset, 10) + 1).toString(),
+              },
+            ]);
+            await consumer.stop();
```
Comment (milindl) [inline test/promisified/admin/fetch_offsets.spec.js:87]: Remove as this has no effect, here and elsewhere

File: test/promisified/admin/fetch_offsets.spec.js [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+          }
+        } catch (error) {
+          if (error.message.includes("Offset out of range")) {
```
Comment (milindl) [inline test/promisified/admin/fetch_offsets.spec.js:90]: I'm not sure when you'll get this error, because it's handled automatically. Maybe remove the try-catch block entirely?

Here, and elsewhere.

File: test/promisified/admin/fetch_offsets.spec.js [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+    producer && (await producer.disconnect());
+    consumer && (await consumer.disconnect());
+  });
+
+  test("should timeout when fetching offsets", async () => {
+    await admin.connect();
+
+    topicName = `test-topic-${secureRandom()}`;
+
+    await createTopic({ topic: topicName, partitions: 1 });
```
[the remark this answers is not in the document]
Comment (milindl) [reply test/promisified/admin/fetch_offsets.spec.js:43]: Make it similar to the other tests. Consumer, producer can also be initialized in beforeEach unless you have to pass them some special parameters. This will make the tests a lot shorter.

File: test/promisified/admin/fetch_offsets.spec.js [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+        ({ leaderEpoch, ...restPartitions }) => restPartitions
+      );
+      return { ...rest, partitions: newPartitions };
+    });
+
+    expect(messagesConsumed.length).toEqual(5);
+    expect(resultWithoutLeaderEpoch).toEqual([
+      {
+        topic: topicName,
+        partitions: [{ partition: 0, offset: 5 }],
```
Comment (milindl) [inline test/promisified/admin/fetch_offsets.spec.js:193]: Wouldn't this be 5+10 instead for this case? or perhaps I've missed something in the test

File: src/common.h [hunk 11]
```diff
@@ -129,7 +129,10 @@ v8::Local<v8::Object> FromDescribeConsumerGroupsResult(
 // DeleteGroups: Response
 v8::Local<v8::Array> FromDeleteGroupsResult(
     const rd_kafka_DeleteGroups_result_t *);
-}  // namespace Admin
+
+v8::Local<v8::Array> FromFetchOffsetsResult(
+    const rd_kafka_ListConsumerGroupOffsets_result_t *result);
+}// namespace Admin
```
Comment (milindl) [inline src/common.h:135]: nit
```suggestion
} // namespace Admin
```

File: src/workers.cc [hunk 12]
```diff
@@ -1424,6 +1424,64 @@ void AdminClientDeleteGroups::HandleErrorCallback() {
   callback->Call(argc, argv);
 }
 
+/**
+ * @brief Fetch Offsets in an asynchronous worker
+ *
+ * This callback will list all the consumer group offsets for the specified
+ * topic offsets.
```
Comment (milindl) [inline src/workers.cc:1431]:
```suggestion
 * groups' topic partitions.
```

**Review by milindl — CHANGES_REQUESTED** [review]
Thanks for the changes! 

Please also add a CHANGELOG.md entry, and merge the master branch to it. Once you do that we will change the base branch of this PR to master.

File: test/promisified/admin/fetch_offsets.spec.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (PratRanj07) [reply test/promisified/admin/fetch_offsets.spec.js:55]: Moved all others except createTopic as we will be creating topics with customised number of partitions for the test

File: examples/kafkajs/fetchOffsets.js [hunk 13]
```diff
@@ -0,0 +1,68 @@
+// require('kafkajs') is replaced with require('@confluentinc/kafka-javascript').KafkaJS.
```
[the remark this answers is not in the document]
Comment (milindl) [reply examples/kafkajs/fetchOffsets.js:1]: Use arguments for bootstrap server, requireStableOffsets and timeout. Use positionals for the group name and the topic/partitions as you are currently doing, it will be something like

```
  const args = parseArgs({
    allowPositionals: true,
    options: {
      'bootstrap-servers': {
        type: 'string',
        short: 'b',
        default: 'localhost:9092',
      },
      'timeout': {
        type: 'string',
        short: 'm',
        default: 5000,
      },
...
    },
  });

and then args.values will contain option values, and args.positionals will contain the rest of the arguments which you can parse with your function that you've already written.

File: src/admin.cc [hunk 14]
```diff
@@ -993,4 +1066,91 @@ NAN_METHOD(AdminClient::NodeDeleteGroups) {
...
+
+  rd_kafka_ListConsumerGroupOffsets_t **requests =
+      static_cast<rd_kafka_ListConsumerGroupOffsets_t **>(
+          malloc(sizeof(rd_kafka_ListConsumerGroupOffsets_t *) *
+                 listGroupOffsets->Length()));
+
+  for (uint32_t i = 0; i < listGroupOffsets->Length(); ++i) {
+    v8::Local<v8::Context> context = Nan::GetCurrentContext();
+    v8::Local<v8::Object> listGroupOffsetObj =
+        listGroupOffsets->Get(context, i).ToLocalChecked().As<v8::Object>();
```
Comment (milindl) [inline src/admin.cc:1097]: Here and at other places, use Nan::Get instead. You don't need to make the context either. Handle the error case properly too, in case the user has not provided a proper object.
```js
    v8::Local<v8::Value> listGroupOffsetValue = Nan::Get(listGroupOffsets, i).ToLocalChecked();
    if (!listGroupOffsetValue->IsObject()) {
      return Nan::ThrowError("Each entry must be an object");
    }
    v8::Local<v8::Object> listGroupOffsetObj = listGroupOffsetValue.As<v8::Object>();
```

File: src/common.cc [hunk 15]
```diff
@@ -440,6 +440,52 @@ std::vector<RdKafka::TopicPartition*> FromV8Array(
...
+    if (!Nan::Get(parameter, i).ToLocal(&v)) {
+      continue;
+    }
+
+    v8::Local<v8::Object> item = v8::Local<v8::Object>::Cast(v);
+
+    v8::Isolate* isolate = v8::Isolate::GetCurrent();
+    v8::Local<v8::Context> context = isolate->GetCurrentContext();
+
+    v8::Local<v8::Value> topicVal =
```
Comment (milindl) [inline src/common.cc:465]: Just use `GetParameter<std::string>` to get the string of the topic name directly, no need for isolate/context. The "item" is an object and thus can be used with `GetParameter<std::string>`. 

Same goes for partitionVal and offset also.

File: src/admin.cc [hunk 16]
```diff
@@ -993,4 +1066,91 @@ NAN_METHOD(AdminClient::NodeDeleteGroups) {
...
+    return Nan::ThrowError("Must provide an array of 'listGroupOffsets'");
+  }
+
+  v8::Local<v8::Array> listGroupOffsets = info[0].As<v8::Array>();
+
+  if (listGroupOffsets->Length() == 0) {
+    return Nan::ThrowError("'listGroupOffsets' cannot be empty");
+  }
+
+  rd_kafka_ListConsumerGroupOffsets_t **requests =
```
Comment (milindl) [inline src/admin.cc:1089]: Add a comment above this telling the reader that the ownership for this is taken by Workers::AdminClientListConsumerGroupOffsets and freeing it is also handled by that class.

File: src/common.cc [hunk 17]
```diff
@@ -1093,6 +1139,117 @@ v8::Local<v8::Array> FromDeleteGroupsResult(
...
+      int32_t leader_epoch =
+          rd_kafka_topic_partition_get_leader_epoch(partition);
+      if (leader_epoch >= 0) {
+        Nan::Set(partition_object, Nan::New("leaderEpoch").ToLocalChecked(),
+                 Nan::New<v8::Number>(leader_epoch));
+      }
+
+      // Set partition-level error (if any)
+      if (partition->err != RD_KAFKA_RESP_ERR_NO_ERROR) {
+        RdKafka::ErrorCode code = static_cast<RdKafka::ErrorCode>(partition->err);
```
Comment (milindl) [inline src/common.cc:1236]: Add // NOLINT or split into two lines.

**Review by milindl — CHANGES_REQUESTED** [review]
Pretty good work to address the review comments. I added a final set of comments, please take a look

File: CHANGELOG.md [hunk 18]
```diff
@@ -1,3 +1,11 @@
+# confluent-kafka-javascript v0.4.0
+
+v0.4.0 is a limited availability feature release. It is supported for all usage.
+
+## Features
+
+1. Add support for an Admin API to list a consumer group's offsets (#49).
+
 # confluent-kafka-javascript v0.3.1
```
Comment (milindl) [inline CHANGELOG.md:9]: 0.3.1 isn't released yet so remove this section and add "1. Fixes an issue where headers were not passed correctly to the `eachBatch` callback (#130)." within this list only. Call it "Enhancements" instead of features.

File: types/kafkajs.d.ts [hunk 19]
```diff
@@ -369,4 +373,11 @@ export type Admin = {
     groups: string[],
     options?: { timeout?: number, includeAuthorizedOperations?: boolean }): Promise<GroupDescriptions>
   deleteGroups(groupIds: string[], options?: { timeout?: number }): Promise<DeleteGroupsResult[]>
+  fetchOffsets(options: { 
+    groupId: string,
+    topics?: TopicInput,  
+    resolveOffsets?: boolean,
```
Comment (milindl) [inline types/kafkajs.d.ts:379]: Sorry I missed this earlier, remove `resolveOffsets` from the types. For users of typescript this makes their life better as they get a compile-time error instead of a run time error.

File: examples/kafkajs/admin/fetchOffsets.js
Comment (milindl) [inline examples/kafkajs/admin/fetchOffsets.js:1]: nit:
rename this file to fetch-offsets.js

**Review by milindl — APPROVED** [review]
Great work. I'm approving the PR. There are some minor changes suggested. You can make them and merge this in.

**Review by milindl — APPROVED** [review]
